include/deal.II/matrix_free/operators.h

```diff
@@ -328,6 +328,75 @@ namespace MatrixFreeOperators
 
 
 
+  /**
+   * This class implements the operation of the action of a Laplace matrix.
+   *
+   * @author Denis Davydov, 2016
+   */
+  template <int dim, int fe_degree, int n_q_points_1d = fe_degree+1, int n_components = 1, typename Number = double>
+  class LaplaceOperator : public Base<dim, Number>
+  {
+  public:
+
+    /**
+     * Constructor.
+     */
+    LaplaceOperator ();
+
+    /**
+     * The diagonal is approximated by computing a local diagonal matrix per element
+     * and distributing it to the global diagonal. This will lead to a slightly
+     * different diagonal for the case with hanging nodes.
+     */
+    virtual void compute_diagonal ();
+
+    /**
+     * Set heterogeneous scalar coefficient @p scalar_coefficient to be used at
+     * the quadrature points. The Table should be of correct size, consistent
+     * with the total number of quadrature points in <code>dim</code>-dimensions,
+     * controlled by @p n_q_points_1d template parameter.
+     */
+    void set_coefficient(const Table<2, VectorizedArray<Number> > &scalar_coefficient );
+
+    virtual void clear();
+
+  private:
+    /**
+     * Applies the laplace matrix operation on an input vector. It is
+     * assumed that the passed input and output vector are correctly initialized
+     * using initialize_dof_vector().
+     */
+    virtual void apply_add (LinearAlgebra::distributed::Vector<Number>       &dst,
+                            const LinearAlgebra::distributed::Vector<Number> &src) const;
+
+    /**
+     * Applies the Laplace operator on a cell.
+     */
+    template <bool homogeneous>
+    void local_apply_cell (const MatrixFree<dim,Number>                     &data,
+                           LinearAlgebra::distributed::Vector<Number>       &dst,
+                           const LinearAlgebra::distributed::Vector<Number> &src,
+                           const std::pair<unsigned int,unsigned int>  &cell_range) const;
+
+    /**
+     * Apply diagonal part of the Laplace operator on a cell.
+     */
+    template <bool hommogeneous>
+    void local_diagonal_cell (const MatrixFree<dim,Number>                &data,
+                              LinearAlgebra::distributed::Vector<Number>  &dst,
+                              const unsigned int &,
+                              const std::pair<unsigned int,unsigned int>  &cell_range) const;
+
+
+    /**
+     * User-provided heterogeneity coefficient.
+     */
+    SmartPointer<const Table<2, VectorizedArray<Number> >,
+                 LaplaceOperator<dim,fe_degree,n_q_points_1d,n_components,Number> > scalar_coefficient;
```

Comment:
I am against `Function` here, because it is not general enough. If someone needs to set up a coefficient from a function -- it's two `for` loops prior to doing `set_coefficient()` (exactly what is done in unit tests).

I must admit I also don't like `std::function` in this context, because we would then need to manage additional (most likely redundant) `Table` inside the class to evaluate coefficient based on a `std::function(cell,q)`. That is because you don't really want to call a wrapper function every time we do `vmult` (contrary to the `projection` where these things are used once only to calculate the RHS).

I am actually in favour of `std::shared_ptr<Table>`. IMHO neither `Function` nor `std::function` are good choices in this context.